### PR TITLE
GEOMESA-3203 Short-circuit disjoint filters in index scans

### PR DESCRIPTION
**Pull request by elahrvivaz**
* Return no ranges for disjoint scans
  * Check for no secondary ranges in tiered scans
* Fix edge cases in filter intersections
* Make test data store scan more realistically

### CODE REVIEW

**Comment by elahrvivaz** [comment]
stacks on top of https://github.com/locationtech/geomesa/pull/2861

File: geomesa-index-api/src/test/scala/org/locationtech/geomesa/index/conf/partition/TablePartitionTest.scala [hunk 1]
```diff
@@ -74,13 +73,19 @@ class TablePartitionTest extends Specification {
...
-      val filter = ECQL.toFilter("dtg < '2018-01-01T00:00:00.000Z' AND dtg > '2018-02-01T00:00:00.000Z'")
-      val partitionOption = TablePartition(ds, sft)
-      partitionOption must beSome
-      val partition = partitionOption.get
-      val partitionsOption = partition.partitions(filter)
-      partitionsOption must beSome
-      partitionsOption.get must beEmpty
+      val filters = Seq(
+        "dtg < '2018-01-01T00:00:00.000Z' AND dtg > '2018-02-01T00:00:00.000Z'",
+        "dtg >= '2014-01-01T12:00:00.000Z' and dtg < '2014-01-01T12:00:00.001Z' and dtg < '2014-01-01T12:00:00.000Z'",
```
Comment (jnh5y) [inline geomesa-index-api/src/test/scala/org/locationtech/geomesa/index/conf/partition/TablePartitionTest.scala:78]: Is `and dtg < '2014-01-01T12:00:00.000Z` repeated on purpose?

File: geomesa-index-api/src/test/scala/org/locationtech/geomesa/index/index/AttributeIndexTest.scala [hunk 2]
```diff
@@ -354,5 +357,23 @@ class AttributeIndexTest extends Specification with LazyLogging {
             features.slice(2, 3)
       }
     }
+
+    "handle filter.EXCLUDE with query hint" in {
```
Comment (jnh5y) [inline geomesa-index-api/src/test/scala/org/locationtech/geomesa/index/index/AttributeIndexTest.scala:361]: Huh, this wasn't working before?

File: geomesa-index-api/src/test/scala/org/locationtech/geomesa/index/conf/partition/TablePartitionTest.scala (hunk 1 above)
Comment (elahrvivaz) [reply geomesa-index-api/src/test/scala/org/locationtech/geomesa/index/conf/partition/TablePartitionTest.scala:78]: the first one is `.001Z` !

File: geomesa-index-api/src/test/scala/org/locationtech/geomesa/index/index/AttributeIndexTest.scala (hunk 2 above)
Comment (elahrvivaz) [reply geomesa-index-api/src/test/scala/org/locationtech/geomesa/index/index/AttributeIndexTest.scala:361]: no, i guess people don't query on Filter.EXCLUDE a lot :P

Comment (jnh5y) [reply geomesa-index-api/src/test/scala/org/locationtech/geomesa/index/index/AttributeIndexTest.scala:361]: Hehehhee.
